### PR TITLE
8333144: docker tests do not work when ubsan is configured

### PR DESCRIPTION
**Pull request by MBaesken**
Currently when we run with ubsan - enabled binaries (configure option --enable-ubsan, see [JDK-8298448](https://bugs.openjdk.org/browse/JDK-8298448)), the docker tests do not work.

We find this in the test output

[STDOUT]
/jdk/bin/java: error while loading shared libraries: libubsan.so.1: cannot open shared object file: No such file or directory

The container where the test is executed does not contain the ubsan package; we might skip the test in this case.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Found leading lowercase letter in issue title for `8333144: docker tests do not work when ubsan is configured`

### Issue
 * [JDK-8333144](https://bugs.openjdk.org/browse/JDK-8333144): docker tests do not work when ubsan is configured (**Bug** - P4)

### Reviewers
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19907/head:pull/19907` \
`$ git checkout pull/19907`

Update a local copy of the PR: \
`$ git checkout pull/19907` \
`$ git pull https://git.openjdk.org/jdk.git pull/19907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19907`

View PR using the GUI difftool: \
`$ git pr show -t 19907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19907.diff">https://git.openjdk.org/jdk/pull/19907.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19907#issuecomment-2191725141)

### CODE REVIEW

**Comment by MBaesken** [comment]
Hello, any input on this ?
Currently the situation with the jtreg tests is not good when running ubsan-enabled binaries .

So we have to check for ubsan (see the PR for an example) or install  ubsan into the docker (or podman or ...) container .

**Review by dholmes-ora — COMMENTED** [review]
On the one hand this seems like a "Dr Dr it hurts when I do this" kind of problem. On the other hand it only affects the docker testing so I'm inclined to let it in, even though it is a bit of a blunt instrument (what if ubsan is installed in the container and someone wants to run with it enabled there?).

**Comment by dholmes-ora** [comment]
I think others who have more investment in this area need to weigh in. I don't know the implications for our infra folk if we need to ensure ubsan is installed.

**Comment by MBaesken** [comment]
> I think others who have more investment in this area need to weigh in. I don't know the implications for our infra folk if we need to ensure ubsan is installed.

I think this would be in the standard container config / BUILDFILE we use for the tests.  So if all works well, no implications.

On the other hand, we could also just run the ubsan - based tests with an own exclude/problem list and exclude the docker test that currently cannot work. That needs a separate list but no other src changes like this PR or the idea with adjusted docker container config.

**Comment by RealCLanger** [comment]
I would prefer to add the required ubsan libraries to the container used for testing when testing an ubsan enabled build. Can we achieve this?

**Comment by MBaesken** [comment]
> I would prefer to add the required ubsan libraries to the container used for testing when testing an ubsan enabled build. Can we achieve this?

I added  libubsan1 to the container  (tested it and works nicely, should do no harm if  we test a non-ubsan build).
Should we go this way ?
If so I could remove the WhiteBox related changes (or keep it for other usages).
I also tried to add a WhiteBox based check to  'DockerTestUtils.java' but this seems not to work. But as i said it is probably not necessary.

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java [hunk 1]
```diff
@@ -30,8 +30,6 @@
  * @modules java.base/jdk.internal.misc
  *          java.management
  *          jdk.jartool/sun.tools.jar
- * @build HelloDocker
- * @run driver DockerBasicTest
```
Comment (RealCLanger) [inline test/hotspot/jtreg/containers/docker/DockerBasicTest.java:34]: you remove @build and @run directives from the test - probably not desired.

**Review by RealCLanger — CHANGES_REQUESTED** [review]
I think adding libubsan1 to the test container is the best way to go. If it cannot be made conditional on ubsan builds then be it so. Then the Whitebox changes should be removed obviously.

**Comment by MBaesken** [comment]
I removed the WhiteBox stuff.
Maybe David could give the change a try in the Oracle CI if that's possible ?

**Review by RealCLanger — APPROVED** [review]
Fine for me.

**Comment by tstuefe** [comment]
I have seen this too late.. some questions:

- Are we sure the images the tests use will always be Debian or Debian descendants? What about RHEL or Oracle Linux?
- How close does the libubsan have to be to the compiler used to build the tested JVM? After all, UBSAN is a compiler feature. Does this work for any version of GCC used to build the JVM? Or do we risk weird test errors if libubsan in the image is incompatible with the GCC version used when building the testee JVM?

**Comment by MBaesken** [comment]
> Are we sure the images the tests use will always be Debian or Debian descendants? What about RHEL or Oracle Linux?

We use Ubuntu for the container test base image as default , see https://github.com/openjdk/jdk/blob/65646b5f81279a7fcef3ea04ef9894cf66f77a5a/test/lib/jdk/test/lib/containers/docker/DockerfileConfig.java#L47
To be more on the safe side (potentially the image can be switched with jdk.test.docker.image.name) we could add a check (and avoid adding the libubsan1 package if it is not ubuntu).

Regarding compatibility, I've seen no issues (and if you compile _without_ ubsan you would not reference the libubsan1 anyway). So it is for some special configuration.

**Comment by dholmes-ora** [comment]
> Maybe David could give the change a try in the Oracle CI if that's possible ?

Testing ...

**Comment by dholmes-ora** [comment]
This had no impact on our tier5 testing where we test container related stuff.

**Comment by MBaesken** [comment]
Hi David, thanks for testing.
Thinking more about it, should I test in ` generateDockerFile ` for  `baseImage`  name containing ubuntu, so that people using a system property to set an own special image do not run into issues with the added `RUN apt-get install libubsan1`     ?

**Comment by tstuefe** [comment]
> > Are we sure the images the tests use will always be Debian or Debian descendants? What about RHEL or Oracle Linux?
> 
> We use Ubuntu for the container test base image as default , see
> 
> https://github.com/openjdk/jdk/blob/65646b5f81279a7fcef3ea04ef9894cf66f77a5a/test/lib/jdk/test/lib/containers/docker/DockerfileConfig.java#L47
> 
> To be more on the safe side (potentially the image can be switched with jdk.test.docker.image.name) we could add a check (and avoid adding the libubsan1 package if it is not ubuntu).
> 
> Regarding compatibility, I've seen no issues (and if you compile _without_ ubsan you would not reference the libubsan1 anyway). So it is for some special configuration.

Okay, thanks Matthias

**Comment by dholmes-ora** [comment]
@MBaesken I know nothing about container setup, sorry.

**Review by tstuefe — APPROVED** [review]
Looks reasonable

**Comment by jerboaa** [comment]
> If the added apt-get call causes issues for some people running the tests, we can easily add an ubuntu distro check in a follow up.

@MBaesken This breaks container tests on non Debian distros. Please add some form of property that needs to be set to install `libubsan1` on the test containers if you need that. It should have no impact on users not needing/using it. Thanks!

**Comment by tstuefe** [comment]
> > If the added apt-get call causes issues for some people running the tests, we can easily add an ubuntu distro check in a follow up.
> 
> 
> 
> @MBaesken This breaks container tests on non Debian distros. Please add some form of property that needs to be set to install `libubsan1` on the test containers if you need that. It should have no impact on users not needing/using it. Thanks!

weird that Oracle reported no errors since I assume they test on what essentially is RHEL

**Comment by jerboaa** [comment]
> > > If the added apt-get call causes issues for some people running the tests, we can easily add an ubuntu distro check in a follow up.
> > 
> > 
> > @MBaesken This breaks container tests on non Debian distros. Please add some form of property that needs to be set to install `libubsan1` on the test containers if you need that. It should have no impact on users not needing/using it. Thanks!
> 
> weird that Oracle reported no errors since I assume they test on what essentially is RHEL

It would depend whether or not skipped tests are flagged or not. The container tests get silently skipped if the container build of the image to test fails. That might be a reason.

**Comment by MBaesken** [comment]
I created
https://bugs.openjdk.org/browse/JDK-8338550
8338550: Make libubsan1 installation in test container optional

**Comment by jerboaa** [comment]
Thanks!

**Pull request closed without merging**